### PR TITLE
Convert loot-core to TS p1

### PR DESCRIPTION
**Pull request by albertogasparin**
Part 1 of the conversion. Mostly renaming js to ts and making sure things make still sense. Added also handy TS ESLint rules. 

In order to support the various .web/.electron/... I ended up adopting `index.d.ts` as pattern to share type definition. Let me know if that makes sense for you too. Right now the function type definition is duplicated, but the solution will be importing from `index.d.ts` and using `const fn: FnDef = () => ...` that way we can keep all variants in sync from a single type file.

Such rewrite however is better done in another PR otherwise we risk confusing git and loosing history (rename + too many changes). Another thing that might do in the next PR is convert all files to ESModules, as things get confusing between CJS exports, ESM default/named and TS adds extra complains.

### CODE REVIEW

File: packages/loot-core/src/platform/server/log/index.d.ts
Comment (j-f1) [inline packages/loot-core/src/platform/server/log/index.d.ts:1]: Should this have exports?

File: packages/loot-core/src/server/encryption.js [hunk 1]
```diff
@@ -1,6 +1,6 @@
 import * as internals from './encryption-internals';
 
-let uuid = require('../platform/uuid/index.electron.js');
+let uuid = require('../platform/uuid/index.electron');
```
Comment (j-f1) [inline packages/loot-core/src/server/encryption.js:3]: Could this be e.g.
```suggestion
let uuid = require('../platform/uuid');
```

File: packages/loot-core/src/shared/schedules.ts [hunk 2]
```diff
@@ -100,7 +100,7 @@ export function getRecurringDescription(config) {
         let strs = [];
 
         let uniqueDays = new Set(patterns.map(p => p.type));
-        let isSameDay = uniqueDays.length === 1 && !uniqueDays.has('day');
+        let isSameDay = uniqueDays.size === 1 && !uniqueDays.has('day');
```
Comment (j-f1) [inline packages/loot-core/src/shared/schedules.ts:103]: Nice catch!

File: packages/loot-core/src/shared/schedules.ts [hunk 3]
```diff
@@ -152,7 +152,12 @@ export function getRecurringDescription(config) {
 }
 
 export function recurConfigToRSchedule(config) {
-  let base = {
+  let base: {
+    start: Date;
+    frequency: string;
+    byHourOfDay: number[];
+    interval?: unknown;
+  } = {
```
Comment (j-f1) [inline packages/loot-core/src/shared/schedules.ts:160]: Worth declaring this as a named type?

File: packages/loot-core/src/shared/util.test.ts [hunk 4]
```diff
@@ -37,31 +37,37 @@ describe('utility functions', () => {
   test('number formatting works with comma-dot format', () => {
     setNumberFormat({ format: 'comma-dot', hideFraction: false });
     let formatter = getNumberFormat().formatter;
+    // @ts-expect-error TS type for format accepts only a number
     expect(formatter.format('1234.56')).toBe('1,234.56');
```
Comment (j-f1) [inline packages/loot-core/src/shared/util.test.ts:41]: hm… maybe port the tests to use number literals instead and throw an error in the API if it’s passed a string to check that that doesn’t happen? (or loosen the API to take strings)

Comment (albertogasparin) [reply packages/loot-core/src/shared/util.test.ts:41]: Well, [the spec says](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat/format) `format()` accepts "a Number, BigInt, or string, to format". Here we are using TS `Intl.NumberFormat` type definition that however does only accept number type (they seem to [not want to change that](https://github.com/microsoft/TypeScript/issues/52124) ). I feel `@ts-expect-error` is ok under these circumstances, as the implementation is correct and it is just the builtin type that is too strict. But happy to change the test: your call.

Comment (j-f1) [reply packages/loot-core/src/shared/util.test.ts:41]: I think the builtin type is reasonable and it makes sense to explicitly handle the coercion to `number` here to ensure that we don’t accidentally pass in a user-entered string which should be parsed using the current number style.

**Review by j-f1 — APPROVED** [review]
This is great, thanks for working on it!

**Pull request merged**